### PR TITLE
Remove Apple SplitDNS in favor of unified split DNS approach

### LINKED ISSUE #2867: remove `on_error` call
This has been brought up by @thomaseizinger before but we don't have any case were we do anything but log the error with the `on_error` callback. If it's a fatal error we use `on_disconnect` and we are already logging errors from connlib.

Until we have any non-fatal error that we want to show the users we probably should remove this callback that only adds boiler plate.

I'm also fine having the API still there but getting rid of the callbacks.

WDYT @jamilbk @roop ?

### LINKED ISSUE #2886: Don't automatically delete connlib logs after upload to GCP
This leads to fragmentation where some logs are in GCP and others saved to a user's phone. When debugging, it's helpful to have them all in one place.

### PR DESCRIPTION
**Pull request by jamilbk**
Screenshot 2023-12-12 at 11 29 43 PM

Found a reliable way to return default system DNS resolvers on iOS and macOS. Even if this method is not perfect, I think it's still worth pursuing because:

* Many administrators will set an upstream resolver in the portal anyway (bypassing client system resolvers)
* It unifies our Split DNS approach across platforms (assuming we can query the default system resolvers on Windows), allowing connlib to intercept all DNS queries on all platforms. This opens the door for some interesting feature possibilities in the area of malicious query blocking. This also makes DNS bugs easier to investigate because there's only one codepath for packets to take. See https://github.com/firezone/firezone/issues/2859

Draft because it needs more testing and I need to figure out the `RustVec<RustString>` type for the Swift -> Rust FFI.

Refs #2713

### CODE REVIEW

File: swift/apple/FirezoneNetworkExtension/DnsResolvers.swift
Comment (jamilbk) [inline swift/apple/FirezoneNetworkExtension/DnsResolvers.swift:1]: I didn't come up with this. It's a wrapper for `libresolv` take from https://stackoverflow.com/questions/31256024/get-dns-server-ip-from-iphone-settings

Alternatively, we could try to implement this in Rust using `libc` but it looks like Rust's libc for ios doesn't expose the `res_state` and such functions so maybe not.

This technique is the same as reading `/etc/resolv.conf`, but on iOS we _can't actually read the file_ so this little hack gets around that.

It seems to be in code published on the App Store, so I'm assuming risk for rejection is low.

File: rust/connlib/clients/android/src/lib.rs [hunk 1]
```diff
@@ -279,29 +247,9 @@ impl Callbacks for CallbackHandler {
...
-                "(Ljava/lang/String;)Z",
-                &[JValue::from(&error)],
-            )
-        })
-    }
-
     fn roll_log_file(&self) -> Option<PathBuf> {
         self.handle.roll_to_new_file().unwrap_or_else(|e| {
-            tracing::debug!("Failed to roll over to new file: {e}");
-            let _ = self.on_error(&Error::LogFileRollError(e));
```
Comment (jamilbk) [inline rust/connlib/clients/android/src/lib.rs]: Literally everywhere we issue the `on_error` callback, we also log an error right before it. Good riddance!

**Comment by conectado** [comment]
I don't think we should remove `onRemoveRoute`, we will need it for #2022 for CIDR resources specifically.

**Comment by jamilbk** [comment]
@conectado You're completely right. Thanks for the catch! Will do some more testing / fixing this morning.

File: rust/connlib/clients/android/src/lib.rs (hunk 1 above)
Comment (ReactorScram) [reply rust/connlib/clients/android/src/lib.rs]: I guess it makes sense if there won't be any errors from connlib that a user could possibly fix. "Check your Wi-Fi / Ethernet" is already signaled by on_disconnect, right?

**Comment by ReactorScram** [comment]
The Windows change is okay. The others I didn't look in depth yet.

File: rust/connlib/clients/android/src/lib.rs (hunk 1 above)
Comment (thomaseizinger) [reply rust/connlib/clients/android/src/lib.rs]: Could we do this in a separate PR perhaps? I didn't look at the changes yet but if possible, splitting refactorings off from features allows for a cleaner history.

Comment (jamilbk) [reply rust/connlib/clients/android/src/lib.rs]: @thomaseizinger Yeah, that's probably a good idea

Comment (jamilbk) [reply rust/connlib/clients/android/src/lib.rs]: @ReactorScram That's correct, `onError` doesn't trigger any user-actionable errors.

Comment (jamilbk) [reply rust/connlib/clients/android/src/lib.rs]: @thomaseizinger Updated. This PR only contains the DNS changes.

**Review by conectado — APPROVED** [review]
Looks good, only thing, we should error out if there's no upstream dns and we didn't get the system's dns.

**Comment by roop** [comment]
@jamilbk The Apple implementation looks good.

Just some thoughts on the strategy:
  - We're getting the default DNS servers at the time of starting the Firezone tunnel. They may not stay as the default through the lifetime of the Firezone tunnel, because of how other VPN apps or DNS provider apps might set the DNS settings (or unset some DNS settings, like when the other VPN (that co-exists with the Firezone tunnel) is turned off). It would be nice if we can get to know when the DNS servers change, but I can't think of a way to do that.
  - When there's another VPN app that has defined match/search domains but is not using the default route, then Firezone will not know of those domains at all
  - There's a comment in [resolv.h](https://opensource.apple.com/source/libresolv/libresolv-25/resolv.h), which has the declaration for `res_9_getservers` used in this PR:
     ```
    /*
     * Important note regarding the BIND-9 API on Mac OS
     * -------------------------------------------------
     * 
     * Mac OS supports a DNS query routing API (see <dns.h>) which is used by
     * most system services to access DNS.  The BIND-9 APIs described here are
     * a lower-level that does not do query routing or search amongst multiple
     * resolver clients.  The results of DNS queries from this API may differ
     * significantly from the results of queries sent to the <dns.h> API.  We
     * strongly encourage developers to use higher-level APIs where possible.
     */
    ```
    The "multiple resolver clients" in the comment is probably the above scenario.

So, I'd say if we're not concerned about the scenario where there are other VPN / DNS apps co-existing with Firezone, then this PR's strategy is fine.

If not, there's  [NEDNSProxyProvider](https://developer.apple.com/documentation/networkextension/nednsproxyprovider), but there the DNS resolution happens in a separate process (apart from the tunnel process), and that gets complicated.

**Comment by jamilbk** [comment]
Yeah no method will be perfect unfortunately. I suspect we always get the interface / dhcp DNS servers with our method.

When these change we should be getting a tunnel stop event and so we reconnect and read them again.

What we could do is poll this function every 30 seconds to pick up any changes

**Comment by thomaseizinger** [comment]
> What we could do is poll this function every 30 seconds to pick up any changes

My 2c: It doesn't sound like much but small bits of "functionality" like this can make things quite complicated over time. People are used to having to restart "things" when something doesn't work. YMMV but for now, I'd rather have a (documented?) workaround (restart the tunnel when you are messing with your DNS settings) than attempting to solve this by reacting to system changes.

The advantage is that a smaller & leaner codebase is much easier to iterate on! :)

**Comment by jamilbk** [comment]
> when you are messing with your DNS settings

If this was the only scenario that caused a DNS change then I agree. There may be other scenarios where DNS settings change and neither us nor the user is aware of, can't think of any right now, but we'll need to watch out for them.

When the actual network changes (WiFi -> Ethernet), the Apple client will tear the tunnel down and reconnect it so we should OK there.

**Comment by roop** [comment]
@jamilbk:

> When these change we should be getting a tunnel stop event and so we reconnect and read them again.

> When the actual network changes (WiFi -> Ethernet), the Apple client will tear the tunnel down and reconnect it so we should OK there.

We do restart connlib when the underlying interface changes (like wifi -> ethernet). We don't tear the tunnel process down, but we stop/start connlib within the same process (we keep the utun device). So, yes, we're okay in these scenarios.

We don't restart connlib when another VPN tunnel (from another VPN app) gets turned on/off.  We don't currently detect that.

> It would be nice if we can get to know when the DNS servers change, but I can't think of a way to do that.

On second thoughts, I think we should be able to do this by looking at the available interfaces when we monitor path changes (like for detecting wifi -> ethernet changes). I'll explore that possibility.

**Pull request merged**